File: src/ledger.py
"""Total Legibility Layer — append-only SQLite event ledger (Core Constraint 1)."""
from __future__ import annotations

import json
import sqlite3
import uuid
from datetime import datetime
from pathlib import Path
from typing import Any

DB_PATH = Path(__file__).parent.parent / "companies" / "ledger.db"


def _connect() -> sqlite3.Connection:
    DB_PATH.parent.mkdir(parents=True, exist_ok=True)
    conn = sqlite3.connect(str(DB_PATH))
    conn.row_factory = sqlite3.Row
    return conn


def init_ledger() -> None:
    """Create the append-only events table if it doesn't exist."""
    with _connect() as conn:
        conn.execute("""
            CREATE TABLE IF NOT EXISTS events (
                id          INTEGER PRIMARY KEY AUTOINCREMENT,
                event_id    TEXT    NOT NULL UNIQUE,
                company_id  TEXT    NOT NULL,
                event_type  TEXT    NOT NULL,
                agent_type  TEXT,
                payload     TEXT    NOT NULL,
                created_at  TEXT    NOT NULL
            )
        """)
        conn.execute("CREATE INDEX IF NOT EXISTS idx_company ON events(company_id)")
        conn.execute("CREATE INDEX IF NOT EXISTS idx_type    ON events(event_type)")
        conn.commit()


def record(
    company_id: str,
    event_type: str,
    payload: dict[str, Any],
    agent_type: str | None = None,
) -> str:
    """Append an immutable event to the ledger. Returns the event_id."""
    event_id = str(uuid.uuid4())
    with _connect() as conn:
        conn.execute(
            """
            INSERT INTO events (event_id, company_id, event_type, agent_type, payload, created_at)
            VALUES (?, ?, ?, ?, ?, ?)
            """,
            (
                event_id,
                company_id,
                event_type,
                agent_type,
                json.dumps(payload),
                datetime.utcnow().isoformat(),
            ),
        )
        conn.commit()
    return event_id


def query(
    company_id: str,
    event_type: str | None = None,
    limit: int = 500,
) -> list[dict[str, Any]]:
    """Read events for a company, newest-last."""
    with _connect() as conn:
        if event_type:
            rows = conn.execute(
                "SELECT * FROM events WHERE company_id=? AND event_type=? ORDER BY id LIMIT ?",
                (company_id, event_type, limit),
            ).fetchall()
        else:
            rows = conn.execute(
                "SELECT * FROM events WHERE company_id=? ORDER BY id LIMIT ?",
                (company_id, limit),
            ).fetchall()
    result = []
    for row in rows:
        entry = dict(row)
        entry["payload"] = json.loads(entry["payload"])
        result.append(entry)
    return result


def query_failures(company_id: str) -> list[dict[str, Any]]:
    """Return all telemetry entries where success=False."""
    rows = query(company_id, event_type="telemetry")
    return [r for r in rows if not r["payload"].get("success", True)]

File: src/ledger.ts
/**
 * Total Legibility Layer — append-only SQLite event ledger.
 * Every agent action, tool call, policy decision, and telemetry event is
 * written here. Nothing is ever deleted or updated.
 *
 * Uses the built-in node:sqlite module (unflagged since Node 22.13).
 */
import { DatabaseSync, type SQLInputValue } from 'node:sqlite';
import fs from 'node:fs';
import path from 'node:path';
import { resolveCompaniesDir } from './paths.js';

const EVENTS_DDL = `
  CREATE TABLE IF NOT EXISTS events (
    id         INTEGER PRIMARY KEY AUTOINCREMENT,
    ts         TEXT    NOT NULL,
    company_id TEXT    NOT NULL,
    event_type TEXT    NOT NULL,
    agent_type TEXT,
    payload    TEXT    NOT NULL
  )
`;

export function initDb(db: DatabaseSync): void {
  db.exec('PRAGMA journal_mode=WAL');
  db.exec(EVENTS_DDL);
}

export interface LedgerRow {
  id: number;
  ts: string;
  company_id: string;
  event_type: string;
  agent_type: string | null;
  payload: unknown;
}

export interface FailureRow {
  ts: string;
  event_type: string;
  agent_type: string | null;
  payload: Record<string, unknown>;
}

export interface Ledger {
  record(company_id: string, event_type: string, payload: unknown, agent_type?: string): void;
  queryEvents(company_id: string, since?: number, until?: number): LedgerRow[];
  tailEvents(company_id: string, since: number, onRow: (row: LedgerRow) => void, signal: AbortSignal): void;
  queryFailures(company_id: string): FailureRow[];
  close(): void;
}

type RawRow = { id: number; ts: string; company_id: string; event_type: string; agent_type: string | null; payload: string };

export function createLedger(dbPath: string): Ledger {
  fs.mkdirSync(path.dirname(dbPath), { recursive: true });
  const db = new DatabaseSync(dbPath);
  initDb(db);

  return {
    record(company_id, event_type, payload, agent_type) {
      db.prepare(
        'INSERT INTO events (ts, company_id, event_type, agent_type, payload) VALUES (?, ?, ?, ?, ?)',
      ).run(
        new Date().toISOString(),
        company_id,
        event_type,
        agent_type ?? null,
        JSON.stringify(payload),
      );
    },

    queryEvents(company_id, since, until) {
      let sql = 'SELECT id, ts, company_id, event_type, agent_type, payload FROM events WHERE company_id = ?';
      const params: SQLInputValue[] = [company_id];
      if (since !== undefined) { sql += ' AND id > ?'; params.push(since); }
      if (until !== undefined) { sql += ' AND id <= ?'; params.push(until); }
      sql += ' ORDER BY id ASC';
      const rows = db.prepare(sql).all(...params) as RawRow[];
      return rows.map(r => ({ ...r, payload: JSON.parse(r.payload) as unknown }));
    },

    tailEvents(company_id, since, onRow, signal) {
      let lastId = since;
      let timer: NodeJS.Timeout | null = null;
      signal.addEventListener('abort', () => {
        if (timer) { clearTimeout(timer); timer = null; }
      });
      const poll = () => {
        timer = null;
        if (signal.aborted) return;
        const rows = db.prepare(
          'SELECT id, ts, company_id, event_type, agent_type, payload FROM events WHERE company_id = ? AND id > ? ORDER BY id ASC',
        ).all(company_id, lastId) as RawRow[];
        for (const r of rows) {
          lastId = r.id;
          onRow({ ...r, payload: JSON.parse(r.payload) as unknown });
        }
        if (!signal.aborted) timer = setTimeout(poll, 500);
      };
      poll();
    },

    queryFailures(company_id) {
      const rows = db.prepare(`
        SELECT ts, event_type, agent_type, payload
        FROM   events
        WHERE  company_id = ?
          AND  (
            event_type LIKE '%.failed'
            OR event_type LIKE '%.error'
            OR event_type = 'human.escalation_required'
            OR json_extract(payload, '$.success') = 0
          )
        ORDER BY id DESC
        LIMIT 20
      `).all(company_id) as Array<{ ts: string; event_type: string; agent_type: string | null; payload: string }>;

      return rows.map(r => ({
        ts: r.ts,
        event_type: r.event_type,
        agent_type: r.agent_type,
        payload: JSON.parse(r.payload) as Record<string, unknown>,
      }));
    },

    close() {
      db.close();
    },
  };
}

// ── Default instance (lazy) — agents call these without a handle ─────────────

export function defaultDbPath(): string {
  return path.join(resolveCompaniesDir(), 'ledger.db');
}

let _default: Ledger | null = null;

function defaultLedger(): Ledger {
  if (!_default) _default = createLedger(defaultDbPath());
  return _default;
}

export function record(
  company_id: string,
  event_type: string,
  payload: unknown,
  agent_type?: string,
): void {
  defaultLedger().record(company_id, event_type, payload, agent_type);
}

export function queryFailures(company_id: string): FailureRow[] {
  return defaultLedger().queryFailures(company_id);
}
